### PR TITLE
feat: Allow specifying Docker build target 

### PR DESCRIPTION
**Pull request by jonife**
#### Which issue(s) does this change fix?

[#2578](https://github.com/aws/aws-sam-cli/issues/2578)

#### Why is this change necessary?
- DockerBuildArgs param is used to pass parameters to the DockerFile at build time. and SAMCLI doesn't support specifying a target for now.

#### How does it address the issue?
- Allow DockerBuildTarget element under Metadata in the DockerFile.

#### What side effects does this change have?
- For tempate.yaml with or without a specified Docker build Target, Docker image would build success without a warning.

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Comment by mgrandis** [comment]
@jonife 
Are you sure `make pr` passed?
The AppVeyor build has failed, can you run it again locally?

File: tests/unit/lib/build_module/test_app_builder.py [hunk 1]
```diff
@@ -739,6 +739,34 @@ def test_can_build_image_function_under_debug(self, mock_os):
...
+        self.builder._build_lambda_image("Name", metadata)
+        self.assertEqual(
+            self.docker_client_mock.api.build.call_args,
+            call(
+                path=ANY,
+                dockerfile="Dockerfile",
+                tag="name:Tag-debug",
+                buildargs={"a": "b", "SAM_BUILD_MODE": "debug"},
+                decode=True,
+                DockerBuildTarget="target"
```
Comment (mgrandis) [inline tests/unit/lib/build_module/test_app_builder.py:767]: `DockerBuildTarget` doesn't seem like the right name here.

**Review by hoffa — APPROVED** [review]

File: tests/unit/lib/build_module/test_app_builder.py [hunk 2]
```diff
@@ -739,6 +739,35 @@ def test_can_build_image_function_under_debug(self, mock_os):
...
+            "DockerBuildArgs": {"a": "b"},
+            "DockerBuildTarget": "stage",
+        }
+
+        self.docker_client_mock.api.build.return_value = []
+
+        result = self.builder._build_lambda_image("Name", metadata)
+
+        self.assertEqual(result, "name:Tag-debug")
+        self.builder._build_lambda_image("Name", metadata)
```
Comment (hoffa) [inline tests/unit/lib/build_module/test_app_builder.py:758]: Is the duplicate call necessary?

Comment (sriram-mv) [reply tests/unit/lib/build_module/test_app_builder.py:758]: I dont think its necessary.

**Review by sriram-mv — APPROVED** [review]
Small nit as pointed out by @hoffa could we address that, other than LGTM

File: tests/unit/lib/build_module/test_app_builder.py (hunk 2 above)
Comment (sriram-mv) [reply tests/unit/lib/build_module/test_app_builder.py:758]: could we address this?

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: tests/unit/lib/build_module/test_app_builder.py (hunk 2 above)
Comment (wchengru) [reply tests/unit/lib/build_module/test_app_builder.py:758]: It also appears in another test case: https://github.com/aws/aws-sam-cli/blob/49851f58d5b64f65d48d127ef9a5fc4eccf96ed9/tests/unit/lib/build_module/test_app_builder.py#L729

**Review by sriram-mv — APPROVED** [review]

**Comment by DanielLoyAugmedix** [comment]
Is this in the documentation anywhere?

**Comment by aksel** [comment]
Are you considering adding this to the docs? Or is there maybe some way to contribute to the docs?

I was about to do some unnecessary work because this was not documented, but lucky for me I searched a bit first :smile:

**Pull request merged**